cpp/src/arrow/util/list_util.cc

```diff
@@ -0,0 +1,349 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdint>
+#include <vector>
+
+#include "arrow/array/array_nested.h"
+#include "arrow/array/builder_nested.h"
+#include "arrow/array/data.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit_run_reader.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/list_util.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/string.h"
+
+namespace arrow::list_util {
+
+namespace internal {
+
+namespace {
+
+using arrow::internal::checked_cast;
+using arrow::internal::ReverseSetBitRunReader;
+using arrow::internal::SetBitRunReader;
+
+/// \pre input.length() > 0 && input.null_count() != input.length()
+/// \param input A LIST_VIEW or LARGE_LIST_VIEW array
+template <typename offset_type>
+int64_t MinViewOffset(const ArraySpan& input) {
+  const uint8_t* validity = input.buffers[0].data;
+  const auto* offsets = input.GetValues<offset_type>(1);
+  const auto* sizes = input.GetValues<offset_type>(2);
+
+  // Make an access to the sizes buffer only when strictly necessary.
+#define MINIMIZE_MIN_VIEW_OFFSET(i)             \
+  auto offset = offsets[i];                     \
+  if (min_offset.has_value()) {                 \
+    if (offset < *min_offset && sizes[i] > 0) { \
+      if (offset == 0) {                        \
+        return 0;                               \
+      }                                         \
+      min_offset = offset;                      \
+    }                                           \
+  } else {                                      \
+    if (sizes[i] > 0) {                         \
+      if (offset == 0) {                        \
+        return 0;                               \
+      }                                         \
+      min_offset = offset;                      \
+    }                                           \
+  }
+
+  std::optional<offset_type> min_offset;
+  if (validity == nullptr) {
+    for (int64_t i = 0; i < input.length; i++) {
+      MINIMIZE_MIN_VIEW_OFFSET(i);
+    }
+  } else {
+    SetBitRunReader reader(validity, input.offset, input.length);
+    while (true) {
+      const auto run = reader.NextRun();
+      if (run.length == 0) {
+        break;
+      }
+      for (int64_t i = run.position; i < run.position + run.length; ++i) {
+        MINIMIZE_MIN_VIEW_OFFSET(i);
+      }
+    }
+  }
+  return min_offset.value_or(0);
```

Comment:
Hmmm in this case(.i.e `min_offset == 0`) what does it means for return 0?

Reply:
If all the list-views are null or empty, the `min_offset` option will be empty and I return `0` which is safe because the `MaxViewEnd` will also be 0 in this case.

There is an opportunity of optimization here that I hadn't noticed before: I can return the option, and if it's empty I don't have to call `MaxViewEnd`.